### DISCUSSION TITLE
working with Parametric Drawings and dynamic blocks

### DISCUSSION THREAD

**Discussion by ikaryss** [mozman/ezdxf#581] [Q&A]
Greetings!  
Suppose I have a parametric drawing with user parameters. How can I access them and change their values with ezdxf? 
![GUID-DA97DEF7-0A59-44D2-B7B7-1D4EB4DD6765](https://user-images.githubusercontent.com/33479999/142780237-e24ef6c0-aa2b-40f5-86bd-cafd681f5ee2.png)
 
I also have a question about dynamic blocks. For example, I set the lengthening operation for my block and now when I change the parameter "L" - the block is lengthened in a specific way. How can I access a specific block and change a specific parameter in it? 
![Снимок экрана 2021-11-22 005259](https://user-images.githubusercontent.com/33479999/142780275-f65cf516-9e6b-4026-8fe6-c5a816e8fbfe.png)
 Thank you!

**Comment by mozman** [comment]
Dynamic blocks are not documented in the DXF reference and therefore not supported by ezdxf.

**Reply by ikaryss** [reply]
And what about parametric drawings? Can I somehow change parameters?

**Reply by ikaryss** [reply]
> And what about parametric drawings? Can I somehow change parameters?

I found that all parameters are in the object section and they all belong to the "AcDbAssocVariable" class. For example like the parameter "d" in the picture below
![Снимок экрана 2021-11-25 143901](https://user-images.githubusercontent.com/33479999/143435157-c8b2fa59-68a7-4a12-9558-02787b3de4fe.png)  

Then I tried to figure out if I could access it directly and wrote the following code  

```python
dwg_file = ezdxf.readfile(r'example.dxf')
rows = [entity.dxfattribs() for entity in dwg_file.objects.query('ACDBASSOCVARIABLE')]
df = pd.DataFrame(rows)
df
``` 
![Снимок экрана 2021-11-25 151614](https://user-images.githubusercontent.com/33479999/143440175-9885e795-e9cd-489a-a788-57f2cfa1e110.png)
  
So it has only 3 attributes, and I cannot access the "d" parameter directly. I would like to know if there is any other ways to do it? thanks in advance for the answer!

**Reply by mozman** [reply]
If you insist on going this route, I'll show you the road to hell:

I don't have your file, so I show this by another file which contains many undocumented DXF entities:

![blocklineargrip](https://user-images.githubusercontent.com/4230966/143455421-8b01aa41-66ab-420e-be4f-2de5e3812561.png)

```Python
from pprint import pprint
import ezdxf
from ezdxf.lldxf.types import dxftag

DIR = ezdxf.options.test_files_path / "AutodeskSamples"

doc = ezdxf.readfile(DIR / "civil_example-imperial.dxf")
block_linear_grip = doc.entitydb.get("700")

# This reports the DXF type: BLOCKLINEARGRIP
print(block_linear_grip.dxftype())
# But the data type is: <class 'ezdxf.entities.dxfentity.DXFTagStorage'> which just stores the DXF tags for writing back.
print(type(block_linear_grip))

# You can access the raw DXF tags by the `xtags` attribute, which is an `ExtendedTags` object
xtags = block_linear_grip.xtags

# DXF entities contain multiple subclass sections, each starting with a tag `(100, subclass name)`, 
# except for the first section.
# Don't mess around with the tags in front of the first subclass!
# Getting a subclass by name returns a `Tags` object which is a subclass of the standard Python `list`:

acdb_eval_expr = xtags.get_subclass("AcDbEvalExpr")
pprint(acdb_eval_expr)

# [DXFTag(100, 'AcDbEvalExpr'), DXFTag(90, 6), DXFTag(98, 33), DXFTag(99, 70)]
# A DXFTag object has two attributes: code, value. 
# Treat the DXFTag objects as immutable data!
# You can append or replace DXF tags:
acdb_eval_expr[3] = dxftag(99, 71)
# Don't modify the first tag: DXFTag(100, 'AcDbEvalExpr')!

##############################################################################
# Changing parameter "d" in your example would look like this:
acdb_assoc_variable_entity = doc.entitydb.get("2A7")
subclass = acdb_assoc_variable_entity.xtags.get_subclass("AcDbAssocVariable")

# Replace value 150 by 165; group code 1 is a string: (but maybe "AcDbCalc:1.0" is the data value of d?)
subclass[3] = dxftag(1, "165")

doc.saveas("modified.dxf")  # and hope AutoCAD will open the file...
```
Without a documentation, it's like poking around in the mud. Good luck!

**Reply by piaoger** [reply]
> If you insist on going this route, I'll show you the road to hell:

Does ODA support dynamic block? I cannot find any helpful infomation after quick search :) @mozman
